src/sparse/KokkosSparse_BlockCrsMatrix.hpp

```diff
@@ -0,0 +1,940 @@
+  ///   but a number of entries that does not fit in int.
+  template<class OffsetType>
+  KOKKOS_INLINE_FUNCTION
+  SparseBlockRowView (const typename MatrixType::values_type& values,
+                 const typename MatrixType::index_type& colidx__,
+                 const ordinal_type& blockDim,
+                 const ordinal_type& count,
+                 const OffsetType& start,
+                 const typename std::enable_if<std::is_integral<OffsetType>::value, int>::type& = 0) :
+    values_ (&values(start*blockDim*blockDim)), colidx_ (&colidx__(start)), blockDim_(blockDim), length (count)
+  {}
+
+  /// \brief Number of entries (i.e. blocks) in the row.
+  ///
+  /// This is a public const field rather than a public const method,
+  /// in order to avoid possible overhead of a method call if the
+  /// compiler is unable to inline that method call.
+  ///
+  /// We assume that rows contain no duplicate entries (i.e., entries
+  /// with the same column index).  Thus, a row may have up to
+  /// A.numCols() entries.  This means that the correct type of
+  /// 'length' is ordinal_type.
+  /// Here, length refers to the number of blocks in a block-row
+  const ordinal_type length;
+
+
+  // Return a pointer offset to full-row i of values_ array; 
+  //   user responsible for indexing into this pointer correctly
+  // Input: i must be the LOCAL row index within this block-row
+  // Output: pointer to values_ array at start of full row with local index i
+  KOKKOS_INLINE_FUNCTION
+  value_type* full_row_in_block_row (const ordinal_type& i) const {
+    return values_+(i*length*blockDim_) ;
+  }
+
+  // Return a pointer offset to local row i of block K of values_ array; 
+  //   user responsible for indexing into this pointer correctly
+  KOKKOS_INLINE_FUNCTION
+  value_type* local_row_in_block (const ordinal_type& K, const ordinal_type& i) const {
+    return (values_+(K*blockDim_ + i*length*blockDim_)) ;
+  }
+
+  // Return the value at a specified block K with local row and col ids (i,j)
+  KOKKOS_INLINE_FUNCTION
+  value_type& local_block_value (const ordinal_type& K, const ordinal_type& i, const ordinal_type& j) const {
+    return values_[K*blockDim_ + i*length*blockDim_ + j];
+  }
+
+  // Return unmanaged 2D strided View wrapping local block K from this block-row
+  KOKKOS_INLINE_FUNCTION
+  block_values_type block(const ordinal_type& K) const {
+    return block_values_type( &(values_[K*blockDim_]), Kokkos::LayoutStride(blockDim_,length*blockDim_,blockDim_,1) );
+  }
+
+
+  KOKKOS_INLINE_FUNCTION
+  ordinal_type findRelBlockOffset ( const ordinal_type idx_to_match, bool is_sorted = false ) const {
+    ordinal_type offset = std::numeric_limits< ordinal_type >::max();
+    for ( ordinal_type blk_offset = 0; blk_offset < length; ++blk_offset ) {
+      ordinal_type idx = colidx_[blk_offset];
+      if ( idx == idx_to_match ) { offset = blk_offset; } // return relative offset
+    }
```

Comment:
Maybe we need to early return or break when it find out the matched index ?

Reply:
Good catch, thanks! Done.